In [ ]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.tree import DecisionTreeClassifier


wine = pd.read_csv('https://bit.ly/wine_csv_data')
data = wine[['alcohol', 'sugar', 'pH']].to_numpy()
target = wine['class'].to_numpy()


train_input, test_input, train_target, test_target = train_test_split(
    data, target, test_size=0.2, random_state=42
)


params = {
    'min_impurity_decrease': np.arange(0.0001, 0.01, 0.0005),
    'max_depth': range(3, 30, 3),
    'min_samples_split': range(2, 100, 10)
}


gs = GridSearchCV(DecisionTreeClassifier(random_state=42), params, n_jobs=-1)
gs.fit(train_input, train_target)


print("젤 좋은 조합:")
for key, value in gs.best_params_.items():
    if isinstance(value, (np.float64, np.float32)):  
        value = float(value)
    print(f"{key}: {value}")
print()


print("최고점수:", round(gs.best_score_, 4))

best_model = gs.best_estimator_
test_score = best_model.score(test_input, test_target)
print("테스트 세트 점수:", round(test_score, 4))


젤 좋은 조합:
max_depth: 21
min_impurity_decrease: 0.0001
min_samples_split: 2

최고점수: 0.8684
테스트 세트 점수: 0.8654


실행 결과 max_dapth: 21, min_impurity_decrease: 0.0001, min_samples: 0.8654일때 점수가 가장 잘 나온다   
하지만 최고 점수가 test set으로 돌렸을 때의 점수보다 높기에 과적합이 일어났을 가능성이 높다